File: Compiler/Ownership/Allocator.py
import Compiler.Ownership.TypeVariableInfo as TypeVariableInfo

class Allocator(object):
    def __init__(self):
        self._nextOwnershipVar = 0
        self._nextGroupVar = 0
        self._nextBorrow = 0

    def __str__(self):
        return "allocator(%s/%s/%s)" % (self._nextOwnershipVar, self._nextGroupVar, self._nextBorrow)

    def __repr__(self) -> str:
        return self.__str__()

    def __eq__(self, other) -> bool:
        if not isinstance(self, Allocator):
            return False
        return self._nextOwnershipVar == other._nextOwnershipVar and self._nextGroupVar == other._nextGroupVar and self._nextBorrow == other._nextBorrow

    def nextOwnershipVar(self):
        n = self._nextOwnershipVar
        self._nextOwnershipVar += 1
        v = TypeVariableInfo.OwnershipVar()
        v.value = n
        return v

    def nextGroupVar(self):
        n = self._nextGroupVar
        self._nextGroupVar += 1
        v = TypeVariableInfo.GroupVar()
        v.value = n
        return v

    def nextBorrowVar(self):
        v = self._nextBorrow
        self._nextBorrow += 1
        return v

    def nextTypeVariableInfo(self):
        tv_info = TypeVariableInfo.TypeVariableInfo()
        tv_info.ownership_var = self.nextOwnershipVar()
        tv_info.group_var = self.nextGroupVar()
        return tv_info

File: Siko/src/siko/ownership/Allocator.rs
use std::fmt::Display;

use crate::siko::ownership::TypeVariableInfo::{GroupTypeVariable, TypeVariableInfo};

use super::{Borrow::BorrowId, TypeVariableInfo::OwnershipTypeVariable};

#[derive(Debug, PartialEq, Eq)]
pub struct Allocator {
    nextOwnershipVar: u32,
    nextGroupVar: u32,
    nextBorrowVar: u32,
}

impl Allocator {
    pub fn new() -> Allocator {
        Allocator {
            nextOwnershipVar: 0,
            nextGroupVar: 0,
            nextBorrowVar: 0,
        }
    }

    pub fn nextOwnershipVar(&mut self) -> OwnershipTypeVariable {
        let v = self.nextOwnershipVar;
        self.nextOwnershipVar += 1;
        OwnershipTypeVariable { value: v }
    }

    pub fn nextGroupVar(&mut self) -> GroupTypeVariable {
        let v = self.nextGroupVar;
        self.nextGroupVar += 1;
        GroupTypeVariable { value: v }
    }

    pub fn nextBorrowVar(&mut self) -> BorrowId {
        let v = self.nextBorrowVar;
        self.nextBorrowVar += 1;
        BorrowId { value: v }
    }

    pub fn nextTypeVariableInfo(&mut self) -> TypeVariableInfo {
        let mut tvInfo = TypeVariableInfo::new();
        tvInfo.owner = self.nextOwnershipVar();
        tvInfo.group = self.nextGroupVar();
        tvInfo
    }
}

impl Display for Allocator {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "allocator({}/{}/{})",
            self.nextOwnershipVar, self.nextGroupVar, self.nextBorrowVar
        )
    }
}
